### PR TITLE
utils: fix extra slash in Redfish default systems url path

### LINKED ISSUE #10441: Redfish OOBM integration is broken in cloudstack
### problem

Redfish OOBM integration is broken in cloudstack

Red fish implementation was done with this PR

https://github.com/apache/cloudstack/pull/4175

### versions

Cloudstack 4.20

### The steps to reproduce the bug

Steps to reproduce the issue 

1. Configure redfish emulator or the mock up server based on your choice on el8 system  
https://hub.docker.com/r/dmtf/redfish-interface-emulator
 https://github.com/DMTF/Redfish-Interface-Emulator/
 https://github.com/DMTF/Redfish-Mockup-Server 
https://www.redhat.com/en/blog/redfish-manage-servers-automatically  

2. I have used the redfish-interface-emulator to test the integration .  

```
Clone the repo https://github.com/DMTF/Redfish-Interface-Emulator/  
pip3 install -r requirements.txt 
mv emulator-config_dynamic_populate.json emulator-config.json 
python emulator.py   
By default it will run on port 5000 
```

3. Disable the following global setting redfish.ignore.ssl and redfish.use.https   

4. Navigate to cloudstack >host  > configure oobm > select the driver as redfish 

Image

5. Enable OOBM  cloudstack >host  > enable oobm 

6. Check the logs, even though the port is configured as 5000 , Cloudstack still sends the request on default 80 port   

```
2025-02-21 06:02:30,485 WARN  [o.a.c.u.r.RedfishClient] (API-Job-Executor-72:[ctx-f9877630, job-661, ctx-06b862cd]) (logid:64cb0d54) Failed to execute HTTP GET request retry attempt 1/2 [URL: http://10.0.35.45/redfish/v1/Systems/] due to exception org.apache.http.conn.HttpHostConnectException: Connect to 10.0.35.45:80 [/10.0.35.45] failed: Connection refused 

  2025-02-21 06:02:32,486 DEBUG [o.a.c.u.r.RedfishClient] (API-Job-Executor-72:[ctx-f9877630, job-661, ctx-06b862cd]) (logid:64cb0d54) HTTP GET request retry attempt 2/2 [URL: http://10.0.35.45/redfish/v1/Systems/].  
```

7. Run the redfish-interface-emulator on port 80

`python emulator.py -port 80 `

8. Check the logs , cloudstack fails to get Systems from the redfish emulator , it looks http://10.0.35.45/redfish/v1/Systems/ (extra slash) and finds no response 

```
2025-02-21 08:19:25,848 WARN  [o.a.c.o.PowerOperationTask] (pool-6-thread-13:[ctx-f95f02a2]) (logid:) Out-of-band management background task operation=STATUS for host ol8.localdomain failed with: Failed to get System ID for host '10.0.35.45' with request 'GET: http://10.0.35.45/redfish/v1/Systems/'. HTTP status code expected '200' but it got '404'. 
```

9.  Using curl you can get the api responses

```
  [root@ol8 Redfish-Interface-Emulator]# curl -s localhost:80/redfish/v1/Systems
{
    "@odata.type": "#ComputerSystemCollection.ComputerSystemCollection",
    "Name": "Computer System Collection",
    "Members@odata.count": 1,
    "Members": [
        {
            "@odata.id": "/redfish/v1/Systems/437XR1138R2"
        }
    ],
    "@odata.context": "/redfish/v1/$metadata#Systems",
    "@odata.id": "/redfish/v1/Systems",
    "@Redfish.Copyright": "Copyright 2014-2016 DMTF. For the full DMTF copyright policy, see http://www.dmtf.org/about/policies/copyright


[root@ol8 ~]# curl s localhost:80/redfish/v1/Systems/437XR1138R2     | jq .Actions'
{
  "#ComputerSystem.Reset": {
    "target": "/redfish/v1/Systems/437XR1138R2/Actions/ComputerSystem.Reset",
    "ResetType@Redfish.AllowableValues": [
      "On",
      "ForceOff",
      "GracefulShutdown",
      "GracefulRestart",
      "ForceRestart",
      "Nmi",
      "ForceOn",
      "PushPowerButton"
    ]
  },
  "Oem": {
    "#Contoso.Reset": {
      "target": "/redfish/v1/Systems/437XR1138R2/Oem/Contoso/Actions/Contoso.Reset"
    }
  }
}
```

### What to do about it?

The integration with the redfish oobm should be fixed as redfish is more secure than ipmi and is widely considered as a replacement for ipmi

https://community.hpe.com/t5/servers-systems-the-right/redfish-a-more-secure-alternative-to-ipmi/ba-p/7044568

### LINKED ISSUE #10742: Redfish OOBM plugin doesn't support custom port
### problem

Redfish OOBM plugin doesn't work with custom ports given, it only works with port 80 or 443 (if the global setting to use https is enabled).

### versions

ACS 4.20+

### The steps to reproduce the bug

1. Create Redfish emulator, run on custom port & port 80
2. Have the global setting to run on https disabled, try to add OOBM config to a host & enable OOBM for that host
3. When oobm config uses custom port, that port isn't used but only port 80 or 443 is used for STATUS checks.

### What to do about it?

Allow custom ports if/when specified to be used; otherwise could default to just port 80 or 443 for Redfish.

### PR DESCRIPTION
**Pull request by yadvr**
This removes extra slash from the base system url in Redfish oobm client utility.

Fixes #10441

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI
- [ ] test (unit or integration test code)

#### Feature/Enhancement Scale

- [x] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [x] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by rohityadavcloud** [comment]
Requesting @kiranchavala for any quick tips, review & tests. Also cc @Pearl1594 @DaanHoogland @andrijapanicsb @steveroles

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Comment by rohityadavcloud** [comment]
@blueorangutan package

File: utils/src/main/java/org/apache/cloudstack/utils/redfish/RedfishClient.java [hunk 1]
```diff
@@ -265,12 +265,12 @@ private String getRequestPathForCommand(RedfishCmdType cmd, String resourceId) {
             if (StringUtils.isBlank(resourceId)) {
                 throw new RedfishException(String.format("Command '%s' requires a valid resource ID '%s'.", cmd, resourceId));
             }
-            return String.format("%s%s", SYSTEMS_URL_PATH, resourceId);
```
Comment (rohityadavcloud) [inline utils/src/main/java/org/apache/cloudstack/utils/redfish/RedfishClient.java:268]: cc @wido @GabrielBrascher Line #263 is where I worry if not having a trailing slash `/` may break any redfish servers/bmcs ?

**Comment by rohityadavcloud** [comment]
@blueorangutan package

File: utils/src/main/java/org/apache/cloudstack/utils/redfish/RedfishClient.java (hunk 1 above)
Comment (DaanHoogland) [reply utils/src/main/java/org/apache/cloudstack/utils/redfish/RedfishClient.java:268]: so line 263 could read
```
return SYSTEMS_URL_PATH + "/";
```
o.r use File.SEPARATOR and/or String.format(). I think we should test it to make sure, but would not be a big issue

**Review by DaanHoogland — APPROVED** [review]
clgtm

File: utils/src/main/java/org/apache/cloudstack/utils/redfish/RedfishClient.java (hunk 1 above)
Comment (rohityadavcloud) [reply utils/src/main/java/org/apache/cloudstack/utils/redfish/RedfishClient.java:268]: No that's not what I meant - if for the basic discovery of the endpoint, a trailing slash isn't passed some systems could fail (hypothetically, I'm just asking). I couldn't return with the `/` as it breaks the base case per Kiran's findings (if I understood them right)

Comment (DaanHoogland) [reply utils/src/main/java/org/apache/cloudstack/utils/redfish/RedfishClient.java:268]: ok, so do you mean we may have to spilt this case out into different ones?

Comment (rohityadavcloud) [reply utils/src/main/java/org/apache/cloudstack/utils/redfish/RedfishClient.java:268]: TBH, I don't know. Usually, I don't think we should require a trailing `/` (most servers should handle this case).

**Review by kiranchavala — COMMENTED** [review]
@rohityadavcloud 

getting the following exception 

2025-03-28 10:48:38,052 ERROR [c.c.a.ApiAsyncJobDispatcher] (API-Job-Executor-32:[ctx-1ed229cb, job-40]) (logid:75fbfd7a) Unexpected exception while executing org.apache.cloudstack.api.command.admin.outofbandmanagement.IssueOutOfBandManagementPowerActionCmd java.lang.IllegalStateException: This is not a JSON Object.


```

[root@ref-trl-8237-k-Mol8-kiran-chavala-mgmt1 ~]# cat /var/log/cloudstack/management/management-server.log  |grep -i "logid:75fbfd7a"
2025-03-28 10:48:38,016 DEBUG [o.a.c.f.j.i.AsyncJobManagerImpl$5] (API-Job-Executor-32:[ctx-1ed229cb, job-40]) (logid:75fbfd7a) Executing AsyncJob {"accountId":2,"cmd":"org.apache.cloudstack.api.command.admin.outofbandmanagement.IssueOutOfBandManagementPowerActionCmd","cmdInfo":"{\"response\":\"json\",\"ctxUserId\":\"2\",\"sessionkey\":\"ACPGChmkdLjBVz9y7ogyTXjzkkU\",\"action\":\"OFF\",\"hostid\":\"f4fe4f90-1e3b-4c5f-974e-e9d48e5e4c52\",\"httpmethod\":\"GET\",\"ctxStartEventId\":\"177\",\"ctxDetails\":\"{\\\"interface com.cloud.host.Host\\\":\\\"f4fe4f90-1e3b-4c5f-974e-e9d48e5e4c52\\\"}\",\"ctxAccountId\":\"2\",\"cmdEventType\":\"HOST.OOBM.ACTION\"}","cmdVersion":0,"completeMsid":null,"created":null,"id":40,"initMsid":32989224370467,"instanceId":1,"instanceType":"Host","lastPolled":null,"lastUpdated":null,"processStatus":0,"removed":null,"result":null,"resultCode":0,"status":"IN_PROGRESS","userId":2,"uuid":"75fbfd7a-84ee-484e-8744-121d82949182"}
2025-03-28 10:48:38,041 WARN  [o.a.c.m.w.WebhookServiceImpl] (API-Job-Executor-32:[ctx-1ed229cb, job-40, ctx-291cbd2e]) (logid:75fbfd7a) Skipping delivering event Event {"description":"{\"details\":\"Host Id: 1 Action: OFF\",\"event\":\"HOST.OOBM.ACTION\",\"status\":\"Completed\"}","eventId":null,"eventType":"HOST.OOBM.ACTION","eventUuid":null,"resourceType":"Host","resourceUUID":null} to any webhook as account ID is missing
2025-03-28 10:48:38,042 WARN  [o.a.c.f.e.EventDistributorImpl] (API-Job-Executor-32:[ctx-1ed229cb, job-40, ctx-291cbd2e]) (logid:75fbfd7a) Failed to publish event [category: ActionEvent, type: HOST.OOBM.ACTION] on bus webhookEventBus
2025-03-28 10:48:38,052 ERROR [c.c.a.ApiAsyncJobDispatcher] (API-Job-Executor-32:[ctx-1ed229cb, job-40]) (logid:75fbfd7a) Unexpected exception while executing org.apache.cloudstack.api.command.admin.outofbandmanagement.IssueOutOfBandManagementPowerActionCmd java.lang.IllegalStateException: This is not a JSON Object.
2025-03-28 10:48:38,053 DEBUG [o.a.c.f.j.i.AsyncJobManagerImpl] (API-Job-Executor-32:[ctx-1ed229cb, job-40]) (logid:75fbfd7a) Complete async job-40, jobStatus: FAILED, resultCode: 530, result: org.apache.cloudstack.api.response.ExceptionResponse/null/{"uuidList":[],"errorcode":"530","errortext":"This is not a JSON Object."}
2025-03-28 10:48:38,054 DEBUG [o.a.c.f.j.i.AsyncJobManagerImpl] (API-Job-Executor-32:[ctx-1ed229cb, job-40]) (logid:75fbfd7a) Publish async job-40 complete on message bus
2025-03-28 10:48:38,054 DEBUG [o.a.c.f.j.i.AsyncJobManagerImpl] (API-Job-Executor-32:[ctx-1ed229cb, job-40]) (logid:75fbfd7a) Wake up jobs related to job-40
2025-03-28 10:48:38,054 DEBUG [o.a.c.f.j.i.AsyncJobManagerImpl] (API-Job-Executor-32:[ctx-1ed229cb, job-40]) (logid:75fbfd7a) Update db status for job-40
2025-03-28 10:48:38,055 DEBUG [o.a.c.f.j.i.AsyncJobManagerImpl] (API-Job-Executor-32:[ctx-1ed229cb, job-40]) (logid:75fbfd7a) Wake up jobs joined with job-40 and disjoin all subjobs created from job- 40
2025-03-28 10:48:38,062 DEBUG [c.c.a.ApiServer] (API-Job-Executor-32:[ctx-1ed229cb, job-40]) (logid:75fbfd7a) Retrieved cmdEventType from job info: HOST.OOBM.ACTION
2025-03-28 10:48:38,064 DEBUG [o.a.c.f.j.i.AsyncJobManagerImpl$5] (API-Job-Executor-32:[ctx-1ed229cb, job-40]) (logid:75fbfd7a) Done executing org.apache.cloudstack.api.command.admin.outofbandmanagement.IssueOutOfBandManagementPowerActionCmd for job-40
2025-03-28 10:48:38,064 INFO  [o.a.c.f.j.i.AsyncJobMonitor] (API-Job-Executor-32:[ctx-1ed229cb, job-40]) (logid:75fbfd7a) Remove job-40 from job monitoring
```

**Review by wido — APPROVED** [review]
CLGTM

Changes look good, thanks!

**Comment by rohityadavcloud** [comment]
@kiranchavala thanks for the logs, I've found and fixed the 'problem' code that read only the first line of the API response & not the content in its entirety due to which the JSON error happened. It's possible that real world refish bmcs don't have newlines in their json but the emulator does. Please help review/test it.

@blueorangutan package

**Comment by rohityadavcloud** [comment]
@blueorangutan test

**Review by kiranchavala — COMMENTED** [review]
@rohityadavcloud 

Found the following issues when testing 


1. If the port is configured as 8000 , Cloudstack sends the request on the default port 80 

Screenshot 2025-04-16 at 4 26 41 PM


Screenshot 2025-04-16 at 4 25 33 PM


2. Cloudstack should send a POST request to control the hardware , currently we are sending only GET request 

Screenshot 2025-04-16 at 2 21 04 PM


Example : 

`[root@ref-trl-8237-k-Mol8-kiran-chavala-kvm1 ~]# curl -s http://10.0.32.140:8000/redfish/v1/Systems/148a3231-209a-49ce-a8a9-b973dbb623c6
`

It supports the following operations 


```
  "Actions": {
        "#ComputerSystem.Reset": {
            "target": "/redfish/v1/Systems/148a3231-209a-49ce-a8a9-b973dbb623c6/Actions/ComputerSystem.Reset",
            "ResetType@Redfish.AllowableValues": [
                "On",
                "ForceOff",
                "GracefulShutdown",
                "GracefulRestart",
                "ForceRestart",
                "Nmi",
                "ForceOn"
            ]
        }
    },
```


`curl -s -d '{"ResetType":"ForceOff"}'      -H "Content-Type: application/json" -X POST      http://$REDFISH_HOST:$REDFISH_PORT/redfish/v1/Systems/$REDFISH_SYSTEM/Actions/ComputerSystem.Reset`


` curl -s -d '{"ResetType":"ForceOn"}'     -H "Content-Type: application/json" -X POST     http://$REDFISH_HOST:$REDFISH_PORT/redfish/v1/Systems/$REDFISH_SYSTEM/Actions/ComputerSystem.Reset`


I have followed the following blogs to test post request 

https://cloudcult.dev/sushy-emulator-redfish-for-the-virtualization-nation/#part-ii-containerized-sushy-tools

https://cloudcult.dev/fishing-for-sushy-with-curl/

**Comment by rohityadavcloud** [comment]
Thanks @kiranchavala the issue #1 you've raised is a known problem as generally redfish is only on 443/https (when the global setting is on, it only works with 443/https). #2 is easily fixable but I don't know if it's a real-world issue or an issue with the emulator.

@wido @GabrielBrascher do you remember or know if the redfish based servers you've, use GET or POST for actions (power operations including status checks)?

**Comment by kiranchavala** [comment]
> Thanks @kiranchavala
> 
> The issue 1 you've raised is a known problem as generally redfish is only on 443/https (when the global setting is on, it only works with 443/https). I've logged this issue for future fix #10742
> 
> 2 is easily fixable but I don't know if it's a real-world issue or an issue with the emulator.
> 
> @wido @GabrielBrascher do you remember or know if the redfish based servers you've, use GET or POST for actions (power operations including status checks)?

thanks @rohityadavcloud 

According to the official redfish API reference guide , post request is used 

pg 59 in 

https://www.dmtf.org/sites/default/files/standards/documents/DSP0266_1.6.0.pdf

**Comment by rohityadavcloud** [comment]
Thanks @kiranchavala I double checked, I think the problem is in your setup. The error message wasn't clear so I've pushed a commit, it was actually sending a POST request but the emulator isn't supporting them. The default emulator mode is static and states here https://github.com/DMTF/Redfish-Interface-Emulator?tab=readme-ov-file#static-emulation-of-a-mockup that `In a static mock, only HTTP GETs will work`.

Now, the error will say this when it happens:

Screenshot 2025-04-17 at 10 39 00 AM

Using the emulator, the dynamic one seems to be supporting POST requests:
```
$ cp emulator-config_dynamic_populate.json emulator-config.json
$ python3 emulator.py -port 80
...
INFO:werkzeug:127.0.0.1 - - [17/Apr/2025 10:40:38] "POST /redfish/v1/Systems/System-1/Actions/ComputerSystem.Reset HTTP/1.1" 200 -
INFO:root:ComputerSystemCollectionAPI init called
INFO:root:ComputerSystemCollectionAPI GET called
...
```

Maybe could you also try/test against the sushy tools?

@blueorangutan package

**Comment by rohityadavcloud** [comment]
Tested with:

1. [DTMF Redfish Emulator](https://github.com/DMTF/Redfish-Interface-Emulator) - with the dynamic profile, it worked except that the power state wouldn't change (something to do with the emulator itself).
2. [Sushy-tools fake emulator](https://github.com/openstack/sushy-tools/tree/master/sushy_tools/emulator) - it just worked out of the box and worked as expect, except for power cycle oobm command which it doesn't support:
```
[2025-04-17 10:56:08,318] DEBUG in main: Request failed with NotSupportedError: Power state PowerCycle is not supported
127.0.0.1 - - [17/Apr/2025 10:56:08] "POST /redfish/v1/Systems/27946b59-9e44-4fa7-8e91-f3527a1ef094/Actions/ComputerSystem.Reset HTTP/1.1" 500 -
in fake done
```

**Comment by rohityadavcloud** [comment]
@blueorangutan test

**Review by kiranchavala — APPROVED** [review]
LGTM, Tested with the following emulators and Cloudstack is sending the correct POST request

https://github.com/DMTF/Redfish-Interface-Emulator

https://github.com/openstack/sushy-tools/tree/master/sushy_tools/emulator

LOGS for DTMF

```
2025-04-17 08:49:24,832 DEBUG [o.a.c.u.r.RedfishClient] (API-Job-Executor-1:[ctx-24843844, job-62, ctx-67948a30]) (logid:6b7a8269) Sending ComputerSystem.Reset Command 'GracefulShutdown' to host '10.0.32.140' with request 'POST http://10.0.32.140/redfish/v1/Systems/System-1/Actions/ComputerSystem.Reset'   

2025-04-17 08:52:24,201 DEBUG [o.a.c.u.r.RedfishClient] (API-Job-Executor-2:[ctx-0054bdcc, job-63, ctx-660040bc]) (logid:85f4142f) Sending ComputerSystem.Reset Command 'On' to host '10.0.32.140' with request 'POST http://10.0.32.140/redfish/v1/Systems/System-1/Actions/ComputerSystem.Reset'   

2025-04-17 08:53:07,955 DEBUG [o.a.c.u.r.RedfishClient] (API-Job-Executor-3:[ctx-94a92e90, job-64, ctx-74bac251]) (logid:47a5a05e) Sending ComputerSystem.Reset Command 'PowerCycle' to host '10.0.32.140' with request 'POST http://10.0.32.140/redfish/v1/Systems/System-1/Actions/ComputerSystem.Reset'   


2025-04-17 08:53:46,461 DEBUG [o.a.c.u.r.RedfishClient] (API-Job-Executor-4:[ctx-fa2d589c, job-65, ctx-b67d16fa]) (logid:56520d5e) Sending ComputerSystem.Reset Command 'ForceRestart' to host '10.0.32.140' with request 'POST http://10.0.32.140/redfish/v1/Systems/System-1/Actions/ComputerSystem.Reset'   

2025-04-17 08:55:59,417 DEBUG [o.a.c.u.r.RedfishClient] (API-Job-Executor-5:[ctx-e0480ddb, job-66, ctx-7fa91049]) (logid:aa398dc4) Sending ComputerSystem.Reset Command 'GracefulShutdown' to host '10.0.32.140' with request 'POST http://10.0.32.140/redfish/v1/Systems/System-1/Actions/ComputerSystem.Reset'  

2025-04-17 08:56:57,601 DEBUG [o.a.c.u.r.RedfishClient] (API-Job-Executor-6:[ctx-310053aa, job-67, ctx-7d3ef831]) (logid:13c86fa5) Retrieved System ID 'System-1' with request 'GET: http://10.0.32.140/redfish/v1/Systems'

2025-04-17 08:56:57,605 DEBUG [o.a.c.u.r.RedfishClient] (API-Job-Executor-6:[ctx-310053aa, job-67, ctx-7d3ef831]) (logid:13c86fa5) Retrieved System power state 'On' with request 'GET: http://10.0.32.140/redfish/v1/Systems/System-1'
```

Sushy tool

```

2025-04-17 09:18:52,658 DEBUG [o.a.c.u.r.RedfishClient] (API-Job-Executor-7:[ctx-4db9f593, job-68, ctx-9d1e43cd]) (logid:c1f7b01c) Sending ComputerSystem.Reset Command 'GracefulShutdown' to host '10.0.32.140' with request 'POST http://10.0.32.140/redfish/v1/Systems/2d75be19-84fe-41a8-9a9e-111c28b0ea9a/Actions/ComputerSystem.Reset'


2025-04-17 09:20:15,604 DEBUG [o.a.c.u.r.RedfishClient] (API-Job-Executor-8:[ctx-4d7c918c, job-69, ctx-38e7cd37]) (logid:5e0c53ad) Sending ComputerSystem.Reset Command 'On' to host '10.0.32.140' with request 'POST http://10.0.32.140/redfish/v1/Systems/2d75be19-84fe-41a8-9a9e-111c28b0ea9a/Actions/ComputerSystem.Reset'


2025-04-17 09:22:42,301 DEBUG [o.a.c.u.r.RedfishClient] (API-Job-Executor-9:[ctx-5d5b161d, job-70, ctx-156bac08]) (logid:9a065e18) Sending ComputerSystem.Reset Command 'PowerCycle' to host '10.0.32.140' with request 'POST http://10.0.32.140/redfish/v1/Systems/2d75be19-84fe-41a8-9a9e-111c28b0ea9a/Actions/ComputerSystem.Reset'

2025-04-17 09:23:38,048 DEBUG [o.a.c.u.r.RedfishClient] (API-Job-Executor-10:[ctx-c1b96157, job-71, ctx-5e030896]) (logid:dadd195f) Sending ComputerSystem.Reset Command 'ForceRestart' to host '10.0.32.140' with request 'POST http://10.0.32.140/redfish/v1/Systems/2d75be19-84fe-41a8-9a9e-111c28b0ea9a/Actions/ComputerSystem.Reset


2025-04-17 09:24:22,849 DEBUG [o.a.c.u.r.RedfishClient] (API-Job-Executor-11:[ctx-e35a4a38, job-72, ctx-8e74fbce]) (logid:34220137) Sending ComputerSystem.Reset Command 'GracefulShutdown' to host '10.0.32.140' with request 'POST http://10.0.32.140/redfish/v1/Systems/2d75be19-84fe-41a8-9a9e-111c28b0ea9a/Actions/ComputerSystem.Reset'

2025-04-17 09:24:48,304 DEBUG [o.a.c.u.r.RedfishClient] (API-Job-Executor-12:[ctx-ee26da42, job-73, ctx-d7ab286c]) (logid:79af02f4) Retrieved System power state 'Off' with request 'GET: http://10.0.32.140/redfish/v1/Systems/2d75be19-84fe-41a8-9a9e-111c28b0ea9a'
```

**Comment by rohityadavcloud** [comment]
Please merge this as/when you may @Pearl1594 @DaanHoogland

**Comment by DaanHoogland** [comment]
> Please merge this as/when you may @Pearl1594 @DaanHoogland Thanks @kiranchavala for the extended testing. There's still a slight chance we have only tested against emulators, that some edge case was missed.

we cannot know/test against all hardware, merging.
